### PR TITLE
upgrade to net6.0 version + add explicit apphost loading for specific platforms

### PR DESCRIPTION
**Pull request by YuliiaKovalova**
## Problem

Fixes: #https://github.com/dotnet/msbuild/issues/9038

MSBuildLocator expects that hostfxr lib is already loaded in the process - that is true for Windows load library, but not for all platforms (macOS, for example).
Some of our customers face an issue like:
```

07/17/2023 15:48:47 Greenwich Mean Time: Error : 9543 :!!Resolve NET Core SDK: Failed
System.DllNotFoundException: Unable to load shared library 'hostfxr' or one of its dependencies. In order to help diagnose loading problems, consider setting the DYLD_PRINT_LIBRARIES environment variable: 
dlopen(/usr/local/share/dotnet/shared/Microsoft.NETCore.App/7.0.9/hostfxr.dylib, 1): image not found
dlopen(/Users/timjroberts/.vscode/extensions/ms-dotnettools.csdevkit-0.2.100-darwin-x64/components/CPS/platforms/darwin-x64/node_modules/@microsoft/visualstudio-projectsystem-buildhost.darwin-x64/hostfxr.dylib, 1): image not found
dlopen(hostfxr.dylib, 1): image not found
dlopen(/usr/local/share/dotnet/shared/Microsoft.NETCore.App/7.0.9/libhostfxr.dylib, 1): image not found
dlopen(/Users/timjroberts/.vscode/extensions/ms-dotnettools.csdevkit-0.2.100-darwin-x64/components/CPS/platforms/darwin-x64/node_modules/@microsoft/visualstudio-projectsystem-buildhost.darwin-x64/libhostfxr.dylib, 1): image not found
dlopen(libhostfxr.dylib, 1): image not found
dlopen(/usr/local/share/dotnet/shared/Microsoft.NETCore.App/7.0.9/hostfxr, 1): image not found
dlopen(/Users/timjroberts/.vscode/extensions/ms-dotnettools.csdevkit-0.2.100-darwin-x64/components/CPS/platforms/darwin-x64/node_modules/@microsoft/visualstudio-projectsystem-buildhost.darwin-x64/hostfxr, 1): image not found
dlopen(hostfxr, 1): image not found
dlopen(/usr/local/share/dotnet/shared/Microsoft.NETCore.App/7.0.9/libhostfxr, 1): image not found
dlopen(/Users/timjroberts/.vscode/extensions/ms-dotnettools.csdevkit-0.2.100-darwin-x64/components/CPS/platforms/darwin-x64/node_modules/@microsoft/visualstudio-projectsystem-buildhost.darwin-x64/libhostfxr, 1): image not found
dlopen(libhostfxr, 1): image not found
```

## Solution

Microsoft.NETCore.DotNetAppHost is a metadata package that loads a platform-specific assembly in runtime.
Specifying reference to it explicitly
`<PackageReference Include="runtime.$(RuntimeIdentifier).Microsoft.NETCore.DotNetAppHost" Version="7.0.9" />`
 helps to copy the content in project/bin folder , so missed library can be resolved.

### CODE REVIEW

File: src/MSBuildLocator/Microsoft.Build.Locator.csproj [hunk 1]
```diff
@@ -11,6 +11,7 @@
     <SignAssembly>true</SignAssembly>
     <AssemblyOriginatorKeyFile>key.snk</AssemblyOriginatorKeyFile>
     <EnableDefaultNoneItems>false</EnableDefaultNoneItems>
+	<RuntimeIdentifier Condition="'$(RuntimeIdentifier)' == ''">$(NETCoreSdkRuntimeIdentifier)</RuntimeIdentifier>
```
Comment (ladipro) [inline src/MSBuildLocator/Microsoft.Build.Locator.csproj:14]: This makes it work when building locally (`NETCoreSdkRuntimeIdentifier` is what the SDK is running on) but it's probably not enough to build a package that can be used everywhere.

Can we easily enumerate the problematic RIDs and package up the native dependency for all of them?

File: src/MSBuildLocator/build/TraverseRuntimeIdentifiers.targets [hunk 2]
```diff
@@ -0,0 +1,31 @@
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Target Name="TraverseRuntimeIdentifiers" AfterTargets="Restore">
+    <PropertyGroup>
+      <RuntimeIdentifiers>
```
Comment (baronfel) [inline src/MSBuildLocator/build/TraverseRuntimeIdentifiers.targets:4]: I'm worried about getting behind as the RID list updates - can this be data-driven in some way? Is there a rule we could follow here?

Comment (YuliiaKovalova) [reply src/MSBuildLocator/build/TraverseRuntimeIdentifiers.targets:4]: I checked the list of available platform-specific packages and included them
image

File: src/MSBuildLocator/build/TraverseRuntimeIdentifiers.targets [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    <PropertyGroup>
+      <RuntimeIdentifierArray>$(RuntimeIdentifiers.Split(';'))</RuntimeIdentifierArray>
+    </PropertyGroup>
+
+    <ItemGroup>
+      <CurrentRuntimeIdentifier Include="$(RuntimeIdentifierArray)" />
+    </ItemGroup>
+
+    <ItemGroup>
+      <PackageReference Include="runtime.%(CurrentRuntimeIdentifier.Identity).Microsoft.NETCore.DotNetAppHost" Version="7.0.9" />
```
Comment (rainersigwald) [inline src/MSBuildLocator/build/TraverseRuntimeIdentifiers.targets:27]: I don't think I understand what this does. Can you elaborate, please?

Comment (YuliiaKovalova) [reply src/MSBuildLocator/build/TraverseRuntimeIdentifiers.targets:27]: I try to resolve this comment https://github.com/microsoft/MSBuildLocator/pull/224#discussion_r1279191978
I list all possible platforms and try to copy what is avaliable locally to the output folder.

$(NETCoreSdkRuntimeIdentifier) usage I liked more.
Not sure if this the right way.

File: src/MSBuildLocator/Microsoft.Build.Locator.csproj (hunk 1 above)
Comment (YuliiaKovalova) [reply src/MSBuildLocator/Microsoft.Build.Locator.csproj]: I wasn't able to enumerate them easily, but listing runtime assemblies works as expected

**Comment by YuliiaKovalova** [comment]
@ladipro , @rainersigwald  is it ok that we have started to generate this folder in output?
image

**Comment by ladipro** [comment]
@YuliiaKovalova this looks expected, the native libraries should not be taking too much extra space.

> How will it be shipped?

It should become part of the Microsoft.Build.Locator package.

File: src/MSBuildLocator/Microsoft.Build.Locator.csproj [hunk 4]
```diff
@@ -25,6 +25,11 @@
     <PackageReference Include="Microsoft.VisualStudio.SDK.EmbedInteropTypes" Version="15.0.36" PrivateAssets="all" />
   </ItemGroup>
 
+  <ItemGroup Condition="'$(TargetFramework)'=='net6.0'">
+	 <!-- These package references help to resolve assemblies in runtime for platforms, where these are not loaded to AppContext by default (e.g MacOS). -->
+    <PackageReference Include="Microsoft.NETCore.DotNetAppHost" Version="7.0.9" />
```
[the remark this answers is not in the document]
Comment (ladipro) [reply src/MSBuildLocator/Microsoft.Build.Locator.csproj]: Presumably the latest hosting library works with older versions of the runtime, as I think for us it is a requirement to be able to locate a .NET 7 MSBuild using MSBuildLocator built against .NET 6. @vitek-karas, can you please confirm?

**Comment by YuliiaKovalova** [comment]
> > Also, is it possible to cover the P/Invoke with a test? Something that would have previously failed on Mac or Linux.
> 
> Actually, it should still be failing because we're currently not making the call to `nethost!get_hostfxr_path`. Or are the current changes enough to fix the issue? I hope I'm not misunderstanding the changes.

It works as expected now because the source issue with the missed lib is resolved.

`
07/17/2023 15:48:47 Greenwich Mean Time: Error : 9543 :!!Resolve NET Core SDK: Failed
System.DllNotFoundException: Unable to load shared library 'hostfxr' or one of its dependencies. In order to help diagnose loading problems, consider setting the DYLD_PRINT_LIBRARIES environment variable: 
dlopen(/usr/local/share/dotnet/shared/Microsoft.NETCore.App/7.0.9/hostfxr.dylib, 1): image not found
dlopen(/Users/timjroberts/.vscode/extensions/ms-dotnettools.csdevkit-0.2.100-darwin-x64/components/CPS/platforms/darwin-x64/node_modules/@microsoft/visualstudio-projectsystem-buildhost.darwin-x64/hostfxr.dylib, 1): image not found
dlopen(hostfxr.dylib, 1): image not found
dlopen(/usr/local/share/dotnet/shared/Microsoft.NETCore.App/7.0.9/libhostfxr.dylib, 1): image not found
dlopen(/Users/timjroberts/.vscode/extensions/ms-dotnettools.csdevkit-0.2.100-darwin-x64/components/CPS/platforms/darwin-x64/node_modules/@microsoft/visualstudio-projectsystem-buildhost.darwin-x64/libhostfxr.dylib, 1): image not found
dlopen(libhostfxr.dylib, 1): image not found
dlopen(/usr/local/share/dotnet/shared/Microsoft.NETCore.App/7.0.9/hostfxr, 1): image not found
dlopen(/Users/timjroberts/.vscode/extensions/ms-dotnettools.csdevkit-0.2.100-darwin-x64/components/CPS/platforms/darwin-x64/node_modules/@microsoft/visualstudio-projectsystem-buildhost.darwin-x64/hostfxr, 1): image not found
dlopen(hostfxr, 1): image not found
dlopen(/usr/local/share/dotnet/shared/Microsoft.NETCore.App/7.0.9/libhostfxr, 1): image not found
dlopen(/Users/timjroberts/.vscode/extensions/ms-dotnettools.csdevkit-0.2.100-darwin-x64/components/CPS/platforms/darwin-x64/node_modules/@microsoft/visualstudio-projectsystem-buildhost.darwin-x64/libhostfxr, 1): image not found
dlopen(libhostfxr, 1): image not found`

There is no need to request the path.

File: src/MSBuildLocator/Microsoft.Build.Locator.csproj (hunk 4 above)
[the remark this answers is not in the document]
Comment (vitek-karas) [reply src/MSBuildLocator/Microsoft.Build.Locator.csproj]: Thet `nethost` (which is the one thing used from this package) is definitely fully backward compatible. There's obviously no 100% forward compatibility forever, but in general we will try really hard to keep it forward compatible for as long as possible.
So in that sense, use the latest version of the package regardless of what TFM you target (just like it's a good idea to use the latest SDK to build your app regardless of the TFM) - and upgrade this often (once 8 ships, upgrade to that and so on).

**Pull request closed without merging**